### PR TITLE
Add pool name regex in TestScopedSpillInjection

### PR DESCRIPTION
**Pull request by duanmeng**
Add a pool name regex in TestScopedSpillInjection to control
the spill injection for target operators.

### CODE REVIEW

File: velox/exec/Spill.h [hunk 1]
```diff
@@ -454,20 +454,26 @@ SpillPartitionIdSet toSpillPartitionIdSet(
...
 
 /// Test utility that returns true if triggered spill is evaluated to happen,
 /// false otherwise.
-bool testingTriggerSpill();
+///
+/// NOTE: It may not be able to retrieve the exact operator name in, eg,
+/// 'GroupingSet' or 'SortBuffer'. So use memory pool name instead and then
+/// tokenize this pool name and check if any of the tokens are present in the
+/// list of spill operator names.
+bool testingTriggerSpill(const std::string& name);
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:476]: Let's always use pool name such as
```
bool testingTriggerSpill(const std::string& pool);
```
Thanks!

File: velox/exec/Spill.h [hunk 2]
```diff
@@ -454,20 +454,26 @@ SpillPartitionIdSet toSpillPartitionIdSet(
...
 /// 'spillPct' is 20 and 'maxInjections' is 10, continuous calls to
-/// testingTriggerSpill() will keep rolling the dice that has a chance of 20%
-/// triggering until 10 triggers have been invoked.
+/// testingTriggerSpill(operatorType()) will keep rolling the dice that has a
+/// chance of 20% triggering until 10 triggers have been invoked.
 class TestScopedSpillInjection {
  public:
   explicit TestScopedSpillInjection(
       int32_t spillPct,
+      std::vector<std::string> operators = {},
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:463]:
```
s/operators/pools/
```

File: velox/exec/Spill.cpp [hunk 3]
```diff
@@ -327,17 +336,41 @@ tsan_atomic<uint32_t>& injectedSpillCount() {
...
+
+  if (!name.empty() && !testingSpillOperators().empty()) {
+    std::unordered_set<std::string> tokens;
+    std::istringstream iss(name);
+    std::string token;
+    while (std::getline(iss, token, '.')) {
+      tokens.insert(token);
+    }
+
+    bool mayNotSpill = true;
```
Comment (xiaoxmeng) [inline velox/exec/Spill.cpp:357]: Let's check pool first? And we pass regular expression instead of list of pool names here? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@duanmeng thanks for the improvement % comments.

File: velox/exec/Spill.cpp [hunk 4]
```diff
@@ -300,44 +300,72 @@ tsan_atomic<uint32_t>& maxSpillInjections() {
...
+} // namespace
 
 TestScopedSpillInjection::TestScopedSpillInjection(
     int32_t spillPct,
+    const std::string& pools,
     uint32_t maxInjections) {
   VELOX_CHECK_EQ(injectedSpillCount(), 0);
   testingSpillPct() = spillPct;
+
+  if (!pools.empty()) {
```
Comment (xiaoxmeng) [inline velox/exec/Spill.cpp:317]: We probably don't need this check? Thanks!

File: velox/exec/Spill.cpp [hunk 5]
```diff
@@ -300,44 +300,72 @@ tsan_atomic<uint32_t>& maxSpillInjections() {
...
+
 tsan_atomic<uint32_t>& injectedSpillCount() {
   static tsan_atomic<uint32_t> injectedCount{0};
   return injectedCount;
 }
 
-bool testingTriggerSpill() {
+bool testingTriggerSpill(const std::string& pool) {
   // Do not evaluate further if trigger is not set.
+  if (!pool.empty() && testingSpillPools().mark_count() > 0) {
```
Comment (xiaoxmeng) [inline velox/exec/Spill.cpp:350]: Can we just RE2::FullMatch as SharedArbitratorTest.cpp?

File: velox/exec/tests/TableWriteTest.cpp [hunk 6]
```diff
@@ -275,7 +275,8 @@ class TableWriteTest : public HiveConnectorTestBase {
           .assertResults(duckDbSql);
     }
     const auto spillDirectory = exec::test::TempDirectoryPath::create();
-    TestScopedSpillInjection scopedSpillInjection(100);
+    TestScopedSpillInjection scopedSpillInjection(
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:278]: We can just leave this as empty? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@duanmeng thanks for the update!

File: velox/exec/Spill.h [hunk 7]
```diff
@@ -452,23 +454,27 @@ SpillPartitionIdSet toSpillPartitionIdSet(
 /// triggered spill.
 /// 'spillPct' indicates the chance of triggering spilling. 100% means spill
 /// will always be triggered.
+/// 'pools' is a regular expression string used to define the specific memory
+/// pools targeted for injecting spilling.
 /// 'maxInjections' indicates the max number of actual triggering. e.g. when
 /// 'spillPct' is 20 and 'maxInjections' is 10, continuous calls to
-/// testingTriggerSpill() will keep rolling the dice that has a chance of 20%
-/// triggering until 10 triggers have been invoked.
+/// testingTriggerSpill(poolName) will keep rolling the dice that has a
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:461]:
```
s/poolName/poolNameRegExpr/
```

File: velox/exec/Spill.h [hunk 8]
```diff
@@ -452,23 +454,27 @@ SpillPartitionIdSet toSpillPartitionIdSet(
...
+/// chance of 20% triggering until 10 triggers have been invoked.
 class TestScopedSpillInjection {
  public:
   explicit TestScopedSpillInjection(
       int32_t spillPct,
+      std::string poolRegExp = ".*",
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:467]:
```
s/std::string/const std::string&/
```

File: velox/exec/Spill.cpp [hunk 9]
```diff
@@ -300,44 +300,59 @@ tsan_atomic<uint32_t>& maxSpillInjections() {
   static tsan_atomic<uint32_t> maxInjections{0};
   return maxInjections;
 }
-} // namespace
 
 tsan_atomic<uint32_t>& testingSpillPct() {
   static tsan_atomic<uint32_t> spillPct{0};
   return spillPct;
 }
+} // namespace
```
Comment (xiaoxmeng) [inline velox/exec/Spill.cpp:308]: Thanks!

File: velox/exec/Spill.cpp [hunk 10]
```diff
@@ -300,44 +300,59 @@ tsan_atomic<uint32_t>& maxSpillInjections() {
...
 
 TestScopedSpillInjection::~TestScopedSpillInjection() {
   testingSpillPct() = 0;
   injectedSpillCount() = 0;
+  testingSpillPoolRegExp() = ".*";
   maxSpillInjections() = 0;
 }
 
+tsan_atomic<std::string>& testingSpillPoolRegExp() {
+  static tsan_atomic<std::string> spillPools{".*"};
```
Comment (xiaoxmeng) [inline velox/exec/Spill.cpp:329]:
```
s/spillPools/spillPoolRegExp/
```

File: velox/exec/Spill.h [hunk 11]
```diff
@@ -452,23 +454,27 @@ SpillPartitionIdSet toSpillPartitionIdSet(
...
 
   ~TestScopedSpillInjection();
 };
 
 /// Test utility that returns true if triggered spill is evaluated to happen,
 /// false otherwise.
-bool testingTriggerSpill();
+bool testingTriggerSpill(const std::string& poolName = "");
 
+tsan_atomic<std::string>& testingSpillPoolRegExp();
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:477]: Does this need to be public? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@duanmeng LGTM. Thanks % minors!

**Review by xiaoxmeng — APPROVED** [review]

**Review by tanjialiang — APPROVED** [review]
Thanks for the improvement

**Pull request closed without merging**